Optimize search efficiency with Zobrist hash

visitorckw opened the pull request:
Implemented Zobrist hash to reduce the number of redundant searches for identical game states.

For example, when the first player plays in cell b2, the search count has been reduced from 66197 to 37078.

While this optimization provides notable performance gains, it comes with the trade-off of increased memory usage due to the need for a hash table. Additionally, collision occurrences cannot be completely eliminated, with the probability estimated at approximately 1/sqrt(HASH_TABLE_SIZE).

Link: https://www.chessprogramming.org/Zobrist_Hashing

visitorckw commented:
Introducing Zobrist hash also facilitates the incorporation of optimization techniques such as IDDFS (Iterative Deepening Depth-First Search) or PVS (Principal Variation Search) in the future. This is because these techniques necessitate the recording of information about previously searched game states.

jouae commented:
I wonder how you got the search count ? since I want to do something for testing performance of the algorithm. I am not familiar with in c and how to test the performance in general ?

visitorckw commented:
I use the following code to simply calculate the number of times the negamax function is called.
```diff
--- a/ttt.c
+++ b/ttt.c
@@ -299,8 +299,11 @@ int get_score(const char *table, char player)
     return score;
 }
 
+long long int search_count = 0;
+
 int negamax(char *table, int depth, char player, int alpha, int beta, int hash)
 {
+    ++search_count;
     if (check_win(table) != ' ')
         return get_score(table, player);
     if (depth == MAX_SEARCH_DEPTH)
@@ -434,7 +437,9 @@ int main()
         if (turn == ai) {
             memset(hash_key_exist, 0, sizeof(hash_key_exist));
             memset(hash_table, 0, sizeof(hash_table));
+            search_count = 0;
             negamax(table, 0, ai, -100000, 100000, 0);
+            printf("Search count: %lld\n", search_count);
         } else {
             draw_board(table);
             int move;
```

jouae commented:
Thank for your answer!

I want to open an issue about test process for those who want to improve the performance if any good idea please leave a comment on it. Thanks again!

jserv commented:
@visitorckw, can you check [shuffle-chess-engine/src/zobrist.c](https://github.com/ArjunBasandrai/shuffle-chess-engine/blob/main/src/zobrist.c)? We may utilize hardware instructions such as `clz` and `popcnt` to accelerate bitwise operations.

visitorckw commented:
Sure, I will do it later.

visitorckw commented:
After reviewing the above code, I believe that the hashing method I am currently using is more suitable for the tic-tac-toe game. This is because tic-tac-toe does not require handling special rules like capturing en passant or castling, as in chess. Each move simply adds a piece without ever removing any pieces, so after each move, the hash value of the next position can be computed with just one xor operation. However, there are a few areas for improvement:
  1. Introducing a linked list to handle collision issues.
  2. Implementing a 64-bit pseudo-random number generator to reduce collision probability.

visitorckw commented:
Another aspect I'd like to discuss is that, despite the low probability, it's still possible for different board to generate the same hash value. This could lead to an unexplored position being mistakenly identified as explored, resulting in the negamax function returning an incorrect value. Many articles suggest that this probability is low enough to be practically negligible. Current AI implementations always make optimal decisions, but once Zobrist hash is introduced to implement a transposition table, the AI's decisions are no longer theoretically perfect.

jserv commented:
> After reviewing the above code, I believe that the hashing method I am currently using is more suitable for the tic-tac-toe game. This is because tic-tac-toe does not require handling special rules like capturing en passant or castling, as in chess. 
> [...]
> Implementing a 64-bit pseudo-random number generator to reduce collision probability.

Once we have a suitable PRNG, the likelihood that different boards will produce the same hash value is sufficiently low to be considered practically insignificant.

visitorckw committed:
Introduce mt19937-64 for improved quality of PNRG

Introduce the mt19937-64 algorithm, a variant of the Mersenne Twister
PRNG, to enhance the quality of generated random numbers.

Link: http://www.math.sci.hiroshima-u.ac.jp/m-mat/MT/VERSIONS/C-LANG/mt19937-64.c

visitorckw committed:
Implement linked list for Hash Table implementation

Introduce the implementation of a linked list data structure to support
the hash table. The linked list code is adapted from the Linux kernel's
implementation.

Link: https://elixir.bootlin.com/linux/v6.7/source/include/linux/list.h

visitorckw committed:
Implement zobrist hash

Zobrist Hashing is a technique used to efficiently compute hash values
for board positions, significantly reducing the computational overhead
of state comparisons in game trees.

visitorckw committed:
Optimize search efficiency with transposition table

Introduce the utilization of Zobrist hashing to implement a
transposition table, enhancing the search efficiency of the algorithm.
The transposition table serves to prevent redundant searches of
identical game states, thereby optimizing the overall search process.

Introduce the utilization of Zobrist hashing to implement a
transposition table, enhancing the search efficiency of the algorithm.
The transposition table serves to prevent redundant searches of
identical game states, thereby optimizing the overall search process.

Experimental results for the first moves at positions a1 and b2
demonstrate a notable reduction in search count:

+----+---------+---------+
|    | Before  | After   |
+----+---------+---------+
| a1 | 13464   | 10479   |
| b2 | 16531   |  9419   |
+----+---------+---------+

visitorckw commented:
Introduce mt19937-64 to provide higher quality PRNG, and incorporate the Linux kernel's linked list to implement the hash table. Utilized Zobrist hash to implement the transposition table, preventing redundant searches of identical game states.

visitorckw commented:
Experimental results for the first moves at positions a1 and b2 demonstrate a notable reduction in search count:

|    | Before  | After   |
|----|---------|---------|
| a1 | 13464   | 10479   |
| b2 | 16531   | 9419   |

jserv commented on mt19937-64.h:
```diff
@@ -0,0 +1,7 @@
+#pragma once
+
+/* initializes mt[NN] with a seed */
+void init_genrand64(unsigned long long seed);
```
Since MT19937 is a well-known PRNG, we can simply name it `mt19937_init`.

jserv commented on mt19937-64.h:
```diff
@@ -0,0 +1,7 @@
+void init_genrand64(unsigned long long seed);
+
+/* generates a random number on [0, 2^64-1]-interval */
+unsigned long long genrand64_int64(void);
```
Similarly, name it `mt19937_rand`.

jserv commented:
Can you address the consideration of the adoption of MT19937 for 64-bit PRNG?

With full support from processor hardware, achieving around 60 Gbits/s with AES, supplemented by intermittent reseeds from RDRAND, is achievable. When implemented purely in software, we can approach similar performance levels for algorithms such as SplitMix64, Xoroshiro128+, and PCG64. Additionally, with the utilization of vector processing capabilities, ChaCha20 offers competitive speeds and outperforms algorithms like the Mersenne Twister.

Reference:
* [The fastest conventional random number generator that can pass Big Crush?](https://lemire.me/blog/2019/03/19/the-fastest-conventional-random-number-generator-that-can-pass-big-crush/)
* [Comparison of hashing functions with small code size](https://github.com/amakukha/minimal_hashes)

visitorckw commented:
It seems like quite an interesting topic :) ! I'm not particularly familiar with PRNGs, and currently, the reason for adopting the mt19937-64 algorithm is simply because of its widespread recognition and extensive usage. However, at present, the PRNG is only used once for initializing the ```zobrist_table``` (generating a total of N * N * 2 pseudo-random numbers), so I believe the actual efficiency of the PRNG has limited impact on the overall program efficiency. Perhaps we could further experiment and compare different PRNG algorithms, but I think this would require some time for research and analysis.

jserv commented:
Thank @visitorckw for the great work!

jserv commented on agents/negamax.c:
```diff
@@ -6,12 +6,15 @@
 static int history_score_sum[N_GRIDS];
 static int history_count[N_GRIDS];
 
+static unsigned long long hash_value;
```
Shall we explicitly use `uint64_t` for portability?

visitorckw replied:
Yes, it would be safer to use uint64_t. I will propose a patch for this.
